src/audio/tdfb/tdfb.c

```diff
@@ -620,168 +617,142 @@ static int tdfb_cmd_switch_set(struct sof_ipc_ctrl_data *cdata, struct tdfb_comp
-static int tdfb_copy(struct comp_dev *dev)
-{
-	struct comp_copy_limits cl;
-	struct comp_buffer *sourceb, *sinkb;
-	struct comp_buffer __sparse_cache *source_c, *sink_c;
-	struct tdfb_comp_data *cd = comp_get_drvdata(dev);
-	int ret = 0;
-	int n;
-
-	comp_dbg(dev, "tdfb_copy()");
+/*
+ * copy and process stream data from source to sink buffers
+ */
 
-	sourceb = list_first_item(&dev->bsource_list, struct comp_buffer,
-				  sink_list);
-	sinkb = list_first_item(&dev->bsink_list, struct comp_buffer,
-				source_list);
+static int tdfb_process(struct processing_module *mod,
+			struct input_stream_buffer *input_buffers, int num_input_buffers,
+			struct output_stream_buffer *output_buffers, int num_output_buffers)
+{
+	struct comp_dev *dev = mod->dev;
+	struct tdfb_comp_data *cd = module_get_private_data(mod);
+	struct audio_stream __sparse_cache *source = input_buffers[0].data;
+	struct audio_stream __sparse_cache *sink = output_buffers[0].data;
+	int frame_count = input_buffers[0].size;
+	int ret;
 
-	source_c = buffer_acquire(sourceb);
-	sink_c = buffer_acquire(sinkb);
+	comp_dbg(dev, "tdfb_process()");
 
 	/* Check for changed configuration */
 	if (comp_is_new_data_blob_available(cd->model_handler)) {
 		cd->config = comp_get_data_blob(cd->model_handler, NULL, NULL);
-		ret = tdfb_setup(cd, source_c->stream.channels, sink_c->stream.channels);
+		ret = tdfb_setup(mod, source->channels, sink->channels);
 		if (ret < 0) {
-			comp_err(dev, "tdfb_copy(), failed FIR setup");
-			goto out;
+			comp_err(dev, "tdfb_process(), failed FIR setup");
+			return ret;
 		}
 	}
 
 	/* Handle enum controls */
 	if (cd->update) {
 		cd->update = false;
-		ret = tdfb_setup(cd, source_c->stream.channels, sink_c->stream.channels);
+		ret = tdfb_setup(mod, source->channels, sink->channels);
 		if (ret < 0) {
-			comp_err(dev, "tdfb_copy(), failed FIR setup");
-			goto out;
+			comp_err(dev, "tdfb_process(), failed FIR setup");
+			return ret;
 		}
 	}
 
-	/* Get source, sink, number of frames etc. to process. */
-	comp_get_copy_limits(source_c, sink_c, &cl);
-
 	/*
 	 * Process only even number of frames with the FIR function. The
 	 * optimized filter function loads the successive input samples from
 	 * internal delay line with a 64 bit load operation.
 	 */
-	cl.frames = MIN(cl.frames, cd->max_frames);
-	if (cl.frames >= 2) {
-		n = (cl.frames >> 1) << 1;
-
-		/* Run the process function */
-		tdfb_process(dev, source_c, sink_c, n,
-			     n * cl.source_frame_bytes,
-			     n * cl.sink_frame_bytes);
-	}
-
-	if (cd->direction_updates && cd->direction_change) {
-		send_get_ctl_ipc(dev);
-		cd->direction_change = false;
-		comp_dbg(dev, "tdfb_dupd %d %d", cd->az_value_estimate, cd->direction.az_slow);
+	frame_count = MIN(frame_count, cd->max_frames) & ~0x1;
+	if (frame_count) {
+		cd->tdfb_func(cd, &input_buffers[0], &output_buffers[0], frame_count);
+		module_update_buffer_position(&input_buffers[0], &output_buffers[0], frame_count);
```

Comment:
any particular reason to use `&input_buffers[0]` instead of just `input_buffers`? Same for output of course

Reply:
I was trying to be pedantic, since they are arrays, but your proposal is less code ---> I will change.

Reply:
@singalsu any ETA ?

Reply:
Maybe this week. This change is easy but changing the module adapter API for other control types is more work.

Reply:
@lyakh The above change is done now.